app/services/authorization/user_permissible_service.rb

```diff
@@ -65,19 +53,37 @@ def allowed_in_any_entity?(permission, entity_class, in_project: nil)
       AllowedInAnyEntityQuery.new(user:, permissions: perms, entity_class:, in_project:).exists?
     end
 
-    def self.permissions_for(action)
-      return [action] if action.is_a?(OpenProject::AccessControl::Permission)
-      return action if action.is_a?(Array) && action.all? { |a| a.is_a?(OpenProject::AccessControl::Permission) }
+    class << self
+      def permissions_for(action) # rubocop:disable Metrics/PerceivedComplexity
+        return [action] if action.is_a?(OpenProject::AccessControl::Permission)
+        return action if action.is_a?(Array) && action.all?(OpenProject::AccessControl::Permission)
+
+        if action.is_a?(Hash)
+          if action[:controller]&.to_s&.starts_with?('/')
+            action = action.dup
```

Comment:
💭 Curious: Why do we `#dup` the action? I don't see it being used anywhere downstream again within the `.permissions_for` method. Is it preventing something?

Reply:
It's copied over from the other implementations of that method. As we don't know where the hash is coming from, it's best practice to either `dup` it or copy explicit keys from it, as we never know how the hash is used outside. Especially in a class method like this.